captum/metrics/_core/infidelity.py

```diff
@@ -472,31 +489,44 @@ def _next_infidelity(current_n_perturb_samples: int) -> Tensor:
         inputs_fwd = torch.repeat_interleave(
             inputs_fwd, current_n_perturb_samples, dim=0
         )
-        inputs_minus_perturb = inputs_fwd - inputs_perturbed_fwd
+        perturbed_fwd_diffs = inputs_fwd - inputs_perturbed_fwd
         attributions_expanded = tuple(
             torch.repeat_interleave(attribution, current_n_perturb_samples, dim=0)
             for attribution in attributions
         )
+
         attributions_times_perturb = tuple(
             (attribution_expanded * perturbation).view(attribution_expanded.size(0), -1)
             for attribution_expanded, perturbation in zip(
                 attributions_expanded, perturbations
             )
         )
 
-        attribution_times_perturb_sums = sum(
-            [
-                torch.sum(attribution_times_perturb, dim=1)
-                for attribution_times_perturb in attributions_times_perturb
-            ]
+        attr_times_perturb_sums = sum(
+            torch.sum(attribution_times_perturb, dim=1)
+            for attribution_times_perturb in attributions_times_perturb
         )
+        attr_times_perturb_sums = cast(Tensor, attr_times_perturb_sums)
 
-        return torch.sum(
-            torch.pow(
-                attribution_times_perturb_sums - inputs_minus_perturb.view(-1), 2
-            ).view(bsz, -1),
-            dim=1,
-        )
+        # reshape as Tensor(bsz, current_n_perturb_samples)
+        attr_times_perturb_sums = attr_times_perturb_sums.view(bsz, -1)
+        perturbed_fwd_diffs = perturbed_fwd_diffs.view(bsz, -1)
+
+        if normalize:
+            # in order to normalize, we have to aggregate the following tensors
+            # to calculate MSE in its polynomial expansion:
+            # (a-b)^2 = a^2 - 2ab + b^2
+            return (
+                attr_times_perturb_sums.pow(2).sum(-1),
+                (attr_times_perturb_sums * perturbed_fwd_diffs).sum(-1),
+                perturbed_fwd_diffs.pow(2).sum(-1),
+            )
+        else:
+            # returns (a-b)^2 if no need to normalize
+            return ((attr_times_perturb_sums - perturbed_fwd_diffs).pow(2).sum(-1),)
```

Comment:
due to the above issue, I keep two ways for aggregation based on `normalize` instead of always using `a^2-2ab+b^2`:
- `(a-b)^2` if not normalize
- `a^2`, `ab`, `b^2` if normalize

This change allows me to pass the tests.
But still worth noting that in older version of pytorch, when normalize, `a^2-2ab+b^2` will lose some precision compared with direct `(a-b)^2`

Reply:
Makes sense, sounds good, this fix looks good to me!